iroha/src/lib.rs

```diff
@@ -19,96 +19,68 @@ pub mod torii;
-/// Type of `Receiver<CommittedBlock>` which should be used for channels of `CommittedBlock` messages.
-pub type CommittedBlockReceiver = Receiver<VersionedCommittedBlock>;
-/// Type of `Sender<AcceptedTransaction>` which should be used for channels of `AcceptedTransaction` messages.
-pub type TransactionSender = Sender<VersionedAcceptedTransaction>;
-/// Type of `Receiver<AcceptedTransaction>` which should be used for channels of
-/// `AcceptedTransaction` messages.
-pub type TransactionReceiver = Receiver<VersionedAcceptedTransaction>;
-/// Type of `Sender<Message>` which should be used for channels of `Message` messages.
-pub type SumeragiMessageSender = Sender<SumeragiVersionedMessage>;
-/// Type of `Receiver<Message>` which should be used for channels of `Message` messages.
-pub type SumeragiMessageReceiver = Receiver<SumeragiVersionedMessage>;
-/// Type of `Sender<Message>` which should be used for channels of `Message` messages.
-pub type BlockSyncMessageSender = Sender<BlockSyncMessage>;
-/// Type of `Receiver<Message>` which should be used for channels of `Message` messages.
-pub type BlockSyncMessageReceiver = Receiver<BlockSyncMessage>;
+/// From book. Check out here:
+/// <https://doc.rust-lang.org/nomicon/phantom-data.html>
+type PhantomInvariant<T> = PhantomData<fn(T) -> T>;
 
 /// Iroha is an [Orchestrator](https://en.wikipedia.org/wiki/Orchestration_%28computing%29) of the
 /// system. It configure, coordinate and manage transactions and queries processing, work of consensus and storage.
-#[derive(Debug)]
-pub struct Iroha {
-    torii: Torii,
-    queue: Arc<RwLock<Queue>>,
-    sumeragi: Arc<RwLock<Sumeragi>>,
-    kura: Arc<RwLock<Kura>>,
-    transactions_receiver: TransactionReceiver,
-    wsv_blocks_receiver: CommittedBlockReceiver,
-    kura_blocks_receiver: CommittedBlockReceiver,
-    sumeragi_message_receiver: SumeragiMessageReceiver,
-    block_sync_message_receiver: BlockSyncMessageReceiver,
-    wsv: Arc<WorldStateView>,
-    block_sync: Arc<RwLock<BlockSynchronizer>>,
-    genesis_network: Option<GenesisNetwork>,
-    telemetry: Option<Receiver<Telemetry>>,
-    config: Configuration,
+#[derive(Debug, Clone, Copy)]
+pub struct Iroha<Q = Queue, S = Sumeragi<Q>, K = Kura, B = BlockSynchronizer<S>>
+where
+    Q: QueueTrait,
+    S: SumeragiTrait,
+    K: KuraTrait,
+    B: BlockSynchronizerTrait,
+{
+    _phantom_data: PhantomInvariant<(Q, S, K, B)>,
 }
 
-impl Iroha {
-    /// Default `Iroha` constructor used to build it based on the provided `Configuration`.
+impl<Q, S, K, B> Iroha<Q, S, K, B>
+where
+    Q: QueueTrait,
+    S: SumeragiTrait,
+    K: KuraTrait,
+    B: BlockSynchronizerTrait,
+{
+    /// Starts iroha with overriding some parts (might be used for testing and mocks)
+    ///
     /// # Errors
-    /// Fails if fails one of the subparts initialization
-    pub fn new(
+    /// Can fail during reading files or initing subparts.
+    pub async fn start_override(
```

Comment:
You don't need a separate function for this - check https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=402550b8cc388b7e051d5998b668ae29